{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Support Vector Regression (SVR) Model for Deconvolution of Human Immune Cell Types\n",
    "### Non-Negative Least Squate Problem  \n",
    "       \n",
    "       I downloaded RNA-Seq data of the 29 immune cell types from https://www.ncbi.nlm.nih.gov/ with the accession number GEO: GSE107011. Data set is an 58311 x 127 dimensional matrix with expression of genes for sample cell types. We get the PBMC values for each sample to generate mixture data, and we find avearge gene expression values for same cell types to generate signature matrix. \n",
    "       Since each sample have different gene expression values, I applied min-max normalization method to mixture and signature data.\n",
    "       After that, I calculate the standard deviation of each gene across cell type in the signature matrix, and  see that many of the gene have similar low variance while some of them have higher variance. Thus, we choose a treshold to generate signature matrix with highest variance genes.\n",
    "       Then, I used SVR model to find the proportion matrix. Since the proportion values should be non-negative and sum of the proportion values for each sample should be one, I made the negative values zero, and normalized to sum up 1.\n",
    "       \n",
    "\n",
    "    \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DZQV_CD8_naive</th>\n",
       "      <th>DZQV_CD8_CM</th>\n",
       "      <th>DZQV_CD8_EM</th>\n",
       "      <th>DZQV_CD8_TE</th>\n",
       "      <th>DZQV_MAIT</th>\n",
       "      <th>DZQV_VD2+</th>\n",
       "      <th>DZQV_VD2-</th>\n",
       "      <th>DZQV_TFH</th>\n",
       "      <th>DZQV_Treg</th>\n",
       "      <th>DZQV_Th1</th>\n",
       "      <th>...</th>\n",
       "      <th>G4YW_mDC</th>\n",
       "      <th>G4YW_Neutrophils</th>\n",
       "      <th>G4YW_Basophils</th>\n",
       "      <th>DZQV_PBMC</th>\n",
       "      <th>925L_PBMC</th>\n",
       "      <th>9JD4_PBMC</th>\n",
       "      <th>G4YW_PBMC</th>\n",
       "      <th>4DUY_PBMC</th>\n",
       "      <th>36TS_PBMC</th>\n",
       "      <th>CR3L_PBMC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>0.214321</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.004214</td>\n",
       "      <td>0.009533</td>\n",
       "      <td>0.011217</td>\n",
       "      <td>0.024504</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.66669</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.248643</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.10965</td>\n",
       "      <td>3.83105</td>\n",
       "      <td>7.432370</td>\n",
       "      <td>3.871090</td>\n",
       "      <td>5.86248</td>\n",
       "      <td>6.943370</td>\n",
       "      <td>1.62892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>4.759630</td>\n",
       "      <td>3.49280</td>\n",
       "      <td>2.648770</td>\n",
       "      <td>3.722320</td>\n",
       "      <td>2.479270</td>\n",
       "      <td>4.188250</td>\n",
       "      <td>3.59954</td>\n",
       "      <td>5.08058</td>\n",
       "      <td>3.20052</td>\n",
       "      <td>8.44928</td>\n",
       "      <td>...</td>\n",
       "      <td>4.65285</td>\n",
       "      <td>1.538540</td>\n",
       "      <td>4.96184</td>\n",
       "      <td>2.06720</td>\n",
       "      <td>3.21687</td>\n",
       "      <td>0.343774</td>\n",
       "      <td>0.783651</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.589092</td>\n",
       "      <td>3.02095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.61518</td>\n",
       "      <td>2.459760</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.586100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.67461</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.85684</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 127 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   DZQV_CD8_naive  DZQV_CD8_CM  DZQV_CD8_EM  DZQV_CD8_TE  \\\n",
       "Unnamed: 0                                                                 \n",
       "ENSG00000223972.5        0.214321      0.00000     0.004214     0.009533   \n",
       "ENSG00000227232.5        4.759630      3.49280     2.648770     3.722320   \n",
       "ENSG00000278267.1        0.000000      2.61518     2.459760     0.000000   \n",
       "ENSG00000243485.5        0.000000      0.00000     0.000000     0.000000   \n",
       "ENSG00000284332.1        0.000000      0.00000     0.000000     0.000000   \n",
       "\n",
       "                   DZQV_MAIT  DZQV_VD2+  DZQV_VD2-  DZQV_TFH  DZQV_Treg  \\\n",
       "Unnamed: 0                                                                \n",
       "ENSG00000223972.5   0.011217   0.024504    0.00000   0.00000    0.66669   \n",
       "ENSG00000227232.5   2.479270   4.188250    3.59954   5.08058    3.20052   \n",
       "ENSG00000278267.1   4.586100   0.000000    0.00000   0.00000    0.00000   \n",
       "ENSG00000243485.5   0.000000   0.000000    0.00000   0.00000    0.00000   \n",
       "ENSG00000284332.1   0.000000   0.000000    0.00000   0.00000    0.00000   \n",
       "\n",
       "                   DZQV_Th1    ...      G4YW_mDC  G4YW_Neutrophils  \\\n",
       "Unnamed: 0                     ...                                   \n",
       "ENSG00000223972.5   0.00000    ...       0.00000          0.248643   \n",
       "ENSG00000227232.5   8.44928    ...       4.65285          1.538540   \n",
       "ENSG00000278267.1   0.00000    ...       1.67461          0.000000   \n",
       "ENSG00000243485.5   0.00000    ...       0.00000          0.000000   \n",
       "ENSG00000284332.1   0.00000    ...       0.00000          0.000000   \n",
       "\n",
       "                   G4YW_Basophils  DZQV_PBMC  925L_PBMC  9JD4_PBMC  G4YW_PBMC  \\\n",
       "Unnamed: 0                                                                      \n",
       "ENSG00000223972.5         0.00000    2.10965    3.83105   7.432370   3.871090   \n",
       "ENSG00000227232.5         4.96184    2.06720    3.21687   0.343774   0.783651   \n",
       "ENSG00000278267.1         0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "ENSG00000243485.5         0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "ENSG00000284332.1         0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "\n",
       "                   4DUY_PBMC  36TS_PBMC  CR3L_PBMC  \n",
       "Unnamed: 0                                          \n",
       "ENSG00000223972.5    5.86248   6.943370    1.62892  \n",
       "ENSG00000227232.5    0.00000   0.589092    3.02095  \n",
       "ENSG00000278267.1    1.85684   0.000000    0.00000  \n",
       "ENSG00000243485.5    0.00000   0.000000    0.00000  \n",
       "ENSG00000284332.1    0.00000   0.000000    0.00000  \n",
       "\n",
       "[5 rows x 127 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pylab as plt\n",
    "import numpy as np\n",
    "from sklearn.svm import NuSVR\n",
    "from sklearn.metrics import r2_score\n",
    "#read the data \n",
    "data =pd.read_csv('CHALLENGE DATA.csv')\n",
    "#set gene names as index\n",
    "data=data.set_index('Unnamed: 0')\n",
    "data.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(58311, 127)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CYFZ_PBMC</th>\n",
       "      <th>FY2H_PBMC</th>\n",
       "      <th>FLWA_PBMC</th>\n",
       "      <th>453W_PBMC</th>\n",
       "      <th>684C_PBMC</th>\n",
       "      <th>CZJE_PBMC</th>\n",
       "      <th>DZQV_PBMC</th>\n",
       "      <th>925L_PBMC</th>\n",
       "      <th>9JD4_PBMC</th>\n",
       "      <th>G4YW_PBMC</th>\n",
       "      <th>4DUY_PBMC</th>\n",
       "      <th>36TS_PBMC</th>\n",
       "      <th>CR3L_PBMC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>3.01149</td>\n",
       "      <td>2.54756</td>\n",
       "      <td>1.86471</td>\n",
       "      <td>2.01926</td>\n",
       "      <td>1.05433</td>\n",
       "      <td>2.65111</td>\n",
       "      <td>2.10965</td>\n",
       "      <td>3.83105</td>\n",
       "      <td>7.432370</td>\n",
       "      <td>3.871090</td>\n",
       "      <td>5.86248</td>\n",
       "      <td>6.943370</td>\n",
       "      <td>1.62892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>2.62153</td>\n",
       "      <td>4.85206</td>\n",
       "      <td>2.78854</td>\n",
       "      <td>3.83599</td>\n",
       "      <td>2.34941</td>\n",
       "      <td>2.78231</td>\n",
       "      <td>2.06720</td>\n",
       "      <td>3.21687</td>\n",
       "      <td>0.343774</td>\n",
       "      <td>0.783651</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.589092</td>\n",
       "      <td>3.02095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.81535</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.85684</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332.1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CYFZ_PBMC  FY2H_PBMC  FLWA_PBMC  453W_PBMC  684C_PBMC  \\\n",
       "Unnamed: 0                                                                 \n",
       "ENSG00000223972.5    3.01149    2.54756    1.86471    2.01926    1.05433   \n",
       "ENSG00000227232.5    2.62153    4.85206    2.78854    3.83599    2.34941   \n",
       "ENSG00000278267.1    0.00000    0.00000    0.00000    0.00000    2.81535   \n",
       "ENSG00000243485.5    0.00000    0.00000    0.00000    0.00000    0.00000   \n",
       "ENSG00000284332.1    0.00000    0.00000    0.00000    0.00000    0.00000   \n",
       "\n",
       "                   CZJE_PBMC  DZQV_PBMC  925L_PBMC  9JD4_PBMC  G4YW_PBMC  \\\n",
       "Unnamed: 0                                                                 \n",
       "ENSG00000223972.5    2.65111    2.10965    3.83105   7.432370   3.871090   \n",
       "ENSG00000227232.5    2.78231    2.06720    3.21687   0.343774   0.783651   \n",
       "ENSG00000278267.1    0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "ENSG00000243485.5    0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "ENSG00000284332.1    0.00000    0.00000    0.00000   0.000000   0.000000   \n",
       "\n",
       "                   4DUY_PBMC  36TS_PBMC  CR3L_PBMC  \n",
       "Unnamed: 0                                          \n",
       "ENSG00000223972.5    5.86248   6.943370    1.62892  \n",
       "ENSG00000227232.5    0.00000   0.589092    3.02095  \n",
       "ENSG00000278267.1    1.85684   0.000000    0.00000  \n",
       "ENSG00000243485.5    0.00000   0.000000    0.00000  \n",
       "ENSG00000284332.1    0.00000   0.000000    0.00000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get the PBMC values for each sample to generate mixture data\n",
    "mixture_data=data.loc[:, data.columns.str.contains('PBMC')]\n",
    "mixture_data=mixture_data.fillna(0)\n",
    "mixture_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(58311, 13)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixture_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD8_naive</th>\n",
       "      <th>CD8_CM</th>\n",
       "      <th>CD8_EM</th>\n",
       "      <th>CD8_TE</th>\n",
       "      <th>MAIT</th>\n",
       "      <th>VD2+</th>\n",
       "      <th>VD2-</th>\n",
       "      <th>TFH</th>\n",
       "      <th>Treg</th>\n",
       "      <th>Th1/Th17</th>\n",
       "      <th>...</th>\n",
       "      <th>C_mono</th>\n",
       "      <th>I_mono</th>\n",
       "      <th>NC_mono</th>\n",
       "      <th>NK</th>\n",
       "      <th>pDC</th>\n",
       "      <th>mDC</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>Basophils</th>\n",
       "      <th>CD4_TE</th>\n",
       "      <th>Th1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>0.203406</td>\n",
       "      <td>0.106600</td>\n",
       "      <td>0.287551</td>\n",
       "      <td>0.002920</td>\n",
       "      <td>0.002804</td>\n",
       "      <td>0.003280</td>\n",
       "      <td>0.000435</td>\n",
       "      <td>0.201344</td>\n",
       "      <td>0.227320</td>\n",
       "      <td>0.001728</td>\n",
       "      <td>...</td>\n",
       "      <td>0.118930</td>\n",
       "      <td>0.245839</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001171</td>\n",
       "      <td>0.226908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.219749</td>\n",
       "      <td>0.259817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>3.566488</td>\n",
       "      <td>3.805285</td>\n",
       "      <td>3.880703</td>\n",
       "      <td>3.420498</td>\n",
       "      <td>1.639900</td>\n",
       "      <td>3.187081</td>\n",
       "      <td>3.332605</td>\n",
       "      <td>3.315260</td>\n",
       "      <td>4.409733</td>\n",
       "      <td>3.184920</td>\n",
       "      <td>...</td>\n",
       "      <td>2.557758</td>\n",
       "      <td>0.963459</td>\n",
       "      <td>3.676425</td>\n",
       "      <td>2.805115</td>\n",
       "      <td>2.882265</td>\n",
       "      <td>3.030297</td>\n",
       "      <td>2.195975</td>\n",
       "      <td>2.731362</td>\n",
       "      <td>2.447345</td>\n",
       "      <td>4.507738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.531282</td>\n",
       "      <td>1.180673</td>\n",
       "      <td>1.077480</td>\n",
       "      <td>0.427700</td>\n",
       "      <td>1.618902</td>\n",
       "      <td>1.009516</td>\n",
       "      <td>1.238800</td>\n",
       "      <td>0.954052</td>\n",
       "      <td>2.439870</td>\n",
       "      <td>1.138515</td>\n",
       "      <td>...</td>\n",
       "      <td>0.238785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.477570</td>\n",
       "      <td>1.511792</td>\n",
       "      <td>0.475460</td>\n",
       "      <td>0.873107</td>\n",
       "      <td>0.508742</td>\n",
       "      <td>0.570323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.505495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.258200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.057947</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CD8_naive    CD8_CM    CD8_EM    CD8_TE      MAIT  \\\n",
       "Unnamed: 0                                                             \n",
       "ENSG00000223972.5   0.203406  0.106600  0.287551  0.002920  0.002804   \n",
       "ENSG00000227232.5   3.566488  3.805285  3.880703  3.420498  1.639900   \n",
       "ENSG00000278267.1   0.531282  1.180673  1.077480  0.427700  1.618902   \n",
       "ENSG00000243485.5   0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSG00000284332.1   0.000000  0.258200  0.000000  0.000000  0.000000   \n",
       "\n",
       "                       VD2+      VD2-       TFH      Treg  Th1/Th17    ...     \\\n",
       "Unnamed: 0                                                             ...      \n",
       "ENSG00000223972.5  0.003280  0.000435  0.201344  0.227320  0.001728    ...      \n",
       "ENSG00000227232.5  3.187081  3.332605  3.315260  4.409733  3.184920    ...      \n",
       "ENSG00000278267.1  1.009516  1.238800  0.954052  2.439870  1.138515    ...      \n",
       "ENSG00000243485.5  0.000000  0.000000  0.000000  0.000000  0.000000    ...      \n",
       "ENSG00000284332.1  0.000000  0.000000  0.000000  0.000000  0.000000    ...      \n",
       "\n",
       "                     C_mono    I_mono   NC_mono        NK       pDC       mDC  \\\n",
       "Unnamed: 0                                                                      \n",
       "ENSG00000223972.5  0.118930  0.245839  0.000000  0.001171  0.226908  0.000000   \n",
       "ENSG00000227232.5  2.557758  0.963459  3.676425  2.805115  2.882265  3.030297   \n",
       "ENSG00000278267.1  0.238785  0.000000  0.477570  1.511792  0.475460  0.873107   \n",
       "ENSG00000243485.5  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSG00000284332.1  0.057947  0.000000  0.115893  0.000000  0.000000  0.000000   \n",
       "\n",
       "                   Neutrophils  Basophils    CD4_TE       Th1  \n",
       "Unnamed: 0                                                     \n",
       "ENSG00000223972.5     0.219749   0.259817  0.000000  0.052075  \n",
       "ENSG00000227232.5     2.195975   2.731362  2.447345  4.507738  \n",
       "ENSG00000278267.1     0.508742   0.570323  0.000000  0.505495  \n",
       "ENSG00000243485.5     0.000000   0.000000  0.000000  0.000000  \n",
       "ENSG00000284332.1     0.000000   0.000000  0.000000  0.000000  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#generate signature matrix\n",
    "#drop the PMBC values\n",
    "signature = data.drop(data.filter(regex='PBMC').columns, axis=1)\n",
    "# find avearge gene expression values for same cell types\n",
    "# then create new columns for avearge value of  cell types  \n",
    "\n",
    "signature['CD8_naive']=signature.loc[:, signature.columns.str.contains('CD8_naive')].mean(axis=1)\n",
    "\n",
    "signature['CD8_CM']=signature.loc[:, signature.columns.str.contains('CD8_CM')].mean(axis=1)\n",
    "\n",
    "signature['CD8_EM']=signature.loc[:, signature.columns.str.contains('CD8_EM')].mean(axis=1)\n",
    "\n",
    "signature['CD8_TE']=signature.loc[:, signature.columns.str.contains('CD8_TE')].mean(axis=1)\n",
    "\n",
    "signature['MAIT']=signature.loc[:, signature.columns.str.contains('MAIT')].mean(axis=1)\n",
    "\n",
    "signature['VD2+']=signature.loc[:, signature.columns.str.contains('VD2+')].mean(axis=1)\n",
    "\n",
    "signature['VD2-']=signature.loc[:, signature.columns.str.contains('VD2-')].mean(axis=1)\n",
    "\n",
    "signature['TFH']=signature.loc[:, signature.columns.str.contains('TFH')].mean(axis=1)\n",
    "\n",
    "signature['Treg']=signature.loc[:, signature.columns.str.contains('Treg')].mean(axis=1)\n",
    "\n",
    "signature['Th1/Th17']=signature.loc[:, signature.columns.str.contains('Th1/Th17')].mean(axis=1)\n",
    "\n",
    "signature['Th17']=signature.loc[:, signature.columns.str.contains('Th17')].mean(axis=1)\n",
    "\n",
    "signature['Th2']=signature.loc[:, signature.columns.str.contains('Th2')].mean(axis=1)\n",
    "\n",
    "signature['CD4_naive']=signature.loc[:, signature.columns.str.contains('CD4_naive')].mean(axis=1)\n",
    "\n",
    "signature['Progenitor']=signature.loc[:, signature.columns.str.contains('Progenitor')].mean(axis=1)\n",
    "\n",
    "signature['B_naive']=signature.loc[:, signature.columns.str.contains('B_naive')].mean(axis=1)\n",
    "\n",
    "signature['B_NSM']=signature.loc[:, signature.columns.str.contains('B_NSM')].mean(axis=1)\n",
    "\n",
    "signature['B_EX']=signature.loc[:, signature.columns.str.contains('B_EX')].mean(axis=1)\n",
    "\n",
    "signature['B_SM']=signature.loc[:, signature.columns.str.contains('B_SM')].mean(axis=1)\n",
    "\n",
    "signature['Plasmablasts']=signature.loc[:, signature.columns.str.contains('Plasmablasts')].mean(axis=1)\n",
    "\n",
    "signature['C_mono']=signature.loc[:, signature.columns.str.contains('C_mono')].mean(axis=1)\n",
    "\n",
    "signature['I_mono']=signature.loc[:, signature.columns.str.contains('I_mono')].mean(axis=1)\n",
    "\n",
    "signature['NC_mono']=signature.loc[:, signature.columns.str.contains('NC_mono')].mean(axis=1)\n",
    "\n",
    "signature['NK']=signature.loc[:, signature.columns.str.contains('NK')].mean(axis=1)\n",
    "\n",
    "signature['pDC']=signature.loc[:, signature.columns.str.contains('pDC')].mean(axis=1)\n",
    "\n",
    "signature['mDC']=signature.loc[:, signature.columns.str.contains('mDC')].mean(axis=1)\n",
    "\n",
    "signature['Neutrophils']=signature.loc[:, signature.columns.str.contains('Neutrophils')].mean(axis=1)\n",
    "\n",
    "\n",
    "signature['Basophils']=signature.loc[:, signature.columns.str.contains('Basophils')].mean(axis=1)\n",
    "\n",
    "signature['CD4_TE']=signature.loc[:, signature.columns.str.contains('CD4_TE')].mean(axis=1)\n",
    "\n",
    "# since there is another cell type name with Th1 we need to drop 'Th1/Th17' column first to choose only Th1 \n",
    "\n",
    "Th1 = signature.drop(signature.filter(regex='Th1/Th17').columns, axis=1)\n",
    "signature['Th1']=Th1.loc[:, Th1.columns.str.contains('Th1')].mean(axis=1)\n",
    "\n",
    "#choose last 29 columns to take only cell type with avearge gene expression values\n",
    "signature_matrix = signature[signature.columns[-29:]].copy()\n",
    "signature_matrix=signature_matrix.fillna(0)\n",
    "signature_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalization (df):\n",
    "    'this is a min-max normalization method'\n",
    "    \n",
    "    Max_each_patient = df.max(axis=0)\n",
    "    Min_each_patient=df.min(axis=0)\n",
    "    Division=Max_each_patient-Min_each_patient \n",
    "    df_normalized = df.subtract(Min_each_patient, axis=1)\n",
    "    df_normalized=df.divide(Division,axis=1)\n",
    "    \n",
    "    return df_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CYFZ_PBMC</th>\n",
       "      <th>FY2H_PBMC</th>\n",
       "      <th>FLWA_PBMC</th>\n",
       "      <th>453W_PBMC</th>\n",
       "      <th>684C_PBMC</th>\n",
       "      <th>CZJE_PBMC</th>\n",
       "      <th>DZQV_PBMC</th>\n",
       "      <th>925L_PBMC</th>\n",
       "      <th>9JD4_PBMC</th>\n",
       "      <th>G4YW_PBMC</th>\n",
       "      <th>4DUY_PBMC</th>\n",
       "      <th>36TS_PBMC</th>\n",
       "      <th>CR3L_PBMC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CYFZ_PBMC  FY2H_PBMC  FLWA_PBMC  453W_PBMC  684C_PBMC  \\\n",
       "Unnamed: 0                                                                 \n",
       "ENSG00000223972.5   0.000076   0.000057   0.000064   0.000075   0.000029   \n",
       "ENSG00000227232.5   0.000066   0.000108   0.000095   0.000143   0.000065   \n",
       "ENSG00000278267.1   0.000000   0.000000   0.000000   0.000000   0.000077   \n",
       "ENSG00000243485.5   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSG00000284332.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "\n",
       "                   CZJE_PBMC  DZQV_PBMC  925L_PBMC  9JD4_PBMC  G4YW_PBMC  \\\n",
       "Unnamed: 0                                                                 \n",
       "ENSG00000223972.5   0.000073   0.000062   0.000127   0.000214   0.000109   \n",
       "ENSG00000227232.5   0.000077   0.000061   0.000107   0.000010   0.000022   \n",
       "ENSG00000278267.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSG00000243485.5   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSG00000284332.1   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "\n",
       "                   4DUY_PBMC  36TS_PBMC  CR3L_PBMC  \n",
       "Unnamed: 0                                          \n",
       "ENSG00000223972.5   0.000200   0.000227   0.000050  \n",
       "ENSG00000227232.5   0.000000   0.000019   0.000093  \n",
       "ENSG00000278267.1   0.000063   0.000000   0.000000  \n",
       "ENSG00000243485.5   0.000000   0.000000   0.000000  \n",
       "ENSG00000284332.1   0.000000   0.000000   0.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find normalized data sets\n",
    "normalized_mixture=normalization(mixture_data)\n",
    "normalized_signature=normalization(signature_matrix)\n",
    "normalized_mixture=normalized_mixture.fillna(0)\n",
    "normalized_mixture.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Nu_SVR(mixture,signature):\n",
    "    'main function to find the proportion matrix'\n",
    "    \"Support Vector Regression \"\n",
    "    \n",
    "    cell_type_data=pd.DataFrame(columns=signature.columns.values.tolist())\n",
    "    for column in mixture: \n",
    "        #support vector regression \n",
    "            regr_nusvr = NuSVR(kernel='linear')\n",
    "            train_x=np.asanyarray(signature)\n",
    "            train_y=np.asanyarray(mixture[column])\n",
    "            model=regr_nusvr.fit (train_x, train_y)\n",
    "            result=model.coef_\n",
    "            #make the negative values zero, and normalized to sum up 1\n",
    "            result=np.where(result<0, 0, result) \n",
    "            normalizer = 1 /  np.sum(result)\n",
    "            estimated_cell_fraction= [x * normalizer for x in result]\n",
    "            cell_type=pd.DataFrame(estimated_cell_fraction,index=[column],columns=signature.columns.values.tolist())\n",
    "            cell_type_data=cell_type_data.append(cell_type)\n",
    "    \n",
    "    return cell_type_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x1a24c23828>]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD8CAYAAACVZ8iyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAEcBJREFUeJzt3X+sX3V9x/Hny14oig6wVMeAektgbiUaxSvqdM5IVHDOLhkmxSVjGwvblGSLWxzEjCjxH9wyjJFNydAwnAPG5tZgHTOyLca5ShEFqlav9Qd3dQIDMWoAi+/98f3Uffv1tnxv+70953vv85F8c8/5nM855/O5Pe2r53zOOd9UFZKk1e1JXTdAktQ9w0CSZBhIkgwDSRKGgSQJw0CShGEgScIwkCRhGEiSgJmuGzDqxBNPrNnZ2a6bIUlT5Y477nigqtYf6vq9C4PZ2Vl27NjRdTMkaaok+cbhrO9lIkmSYSBJMgwkSRgGkiQMA0kShoEkCcNAkoRhIEm98MFPfY1b7trT2f4NA0nqges//Q1u3fntzvZvGEiSDANJkmEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAknqhOt6/YSBJPZEO920YSJIMA0mSYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgSb1Q1e3biQwDSeqJdPhyIsNAkmQYSJLGDIMk5ybZlWQ+yaWLLF+b5Ma2fHuS2VZ+VJLrktyd5ItJLpts8yVJk/CEYZBkDXA1cB6wCbggyaaRahcBD1XV6cBVwJWt/A3A2qp6DvAC4Hf3BYUkqT/GOTM4G5ivqt1V9RhwA7B5pM5m4Lo2fTNwTpIw+PKeY5PMAE8GHgO+O5GWS5ImZpwwOBm4d2h+oZUtWqeq9gIPA+sYBMP3gW8B3wT+vKoePMw2S5ImbJwwWOxmp9EbYg9U52zgceBngI3AHyU57Sd2kFycZEeSHffff/8YTZIkTdI4YbAAnDo0fwqw50B12iWh44AHgTcC/1JVP6yq+4BPAXOjO6iqa6pqrqrm1q9fv/ReSJIOyzhhcDtwRpKNSY4GtgBbR+psBS5s0+cDt9XgcbpvAq/MwLHAi4EvTabpkqRJecIwaGMAlwC3Al8EbqqqnUmuSPL6Vu1aYF2SeeAtwL7bT68GngrcwyBUPlhVd024D5I09bp9GQXMjFOpqrYB20bKLh+afoTBbaSj631vsXJJ0k/q8G0UPoEsSTIMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCSpF6rj91EYBpLUE4PvBOuGYSBJMgwkSYaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSeqFotv3URgGktQT3b2MwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkqReq21cTjRcGSc5NsivJfJJLF1m+NsmNbfn2JLNDy56b5NNJdia5O8kxk2u+JK0gHb6c6AnDIMka4GrgPGATcEGSTSPVLgIeqqrTgauAK9u6M8CHgN+rqjOBVwA/nFjrJUkTMc6ZwdnAfFXtrqrHgBuAzSN1NgPXtembgXOSBHg1cFdVfR6gqv63qh6fTNMlSZMyThicDNw7NL/QyhatU1V7gYeBdcDPApXk1iSfTfLWw2+yJGnSZsaos9hVrNGhjgPVmQFeBrwQ+AHwiSR3VNUn9ls5uRi4GGDDhg1jNEmSNEnjnBksAKcOzZ8C7DlQnTZOcBzwYCv/j6p6oKp+AGwDzhrdQVVdU1VzVTW3fv36pfdCknRYxgmD24EzkmxMcjSwBdg6UmcrcGGbPh+4raoKuBV4bpKntJD4JeALk2m6JGlSnvAyUVXtTXIJg3/Y1wAfqKqdSa4AdlTVVuBa4Pok8wzOCLa0dR9K8hcMAqWAbVX10WXqiyTpEI0zZkBVbWNwiWe47PKh6UeANxxg3Q8xuL1UktRTPoEsSTIMJKkPpuJ1FJKk5ZcO30dhGEiSDANJkmEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEm9ke7eRmEYSJIMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJPVCVXW6f8NAknqiw1cTGQaS1AfdnhcYBpLUG761VJJWuY6HDAwDSeqLdDhqYBhIUg9Ux6MGhoEk9YRjBpKkTo0VBknOTbIryXySSxdZvjbJjW359iSzI8s3JPlekj+eTLMlaWXp/QBykjXA1cB5wCbggiSbRqpdBDxUVacDVwFXjiy/CvjY4TdXklauvl8mOhuYr6rdVfUYcAOweaTOZuC6Nn0zcE4y6FaSXwV2Azsn02RJWnmm4aGzk4F7h+YXWtmidapqL/AwsC7JscCfAO84/KZK0krX71tLF2vdaIgdqM47gKuq6nsH3UFycZIdSXbcf//9YzRJklaWrscMZsaoswCcOjR/CrDnAHUWkswAxwEPAi8Czk/yLuB44EdJHqmq9w6vXFXXANcAzM3NdX22JEmd6HLMYJwwuB04I8lG4L+BLcAbR+psBS4EPg2cD9xWg/ex/uK+CkneDnxvNAgkSdD1qMEThkFV7U1yCXArsAb4QFXtTHIFsKOqtgLXAtcnmWdwRrBlORstSStRl6+wHufMgKraBmwbKbt8aPoR4A1PsI23H0L7JElHgE8gS1IPdD2AbBhIUk/0/aEzSdIy6/o2SsNAknrC7zOQpFWuOh40MAwkqSccM5CkVc4xA0kS0O1DZ4aBJMkwkKQ+8KEzSRIA6XAE2TCQpB7w1lJJUucMA0nqAW8tlSQBPnQmSfJuIkkS+KI6SVLHDANJ6gEHkCVJgAPIkrTq+dCZJAnwraWStOo5ZiBJAhwzkKRVz1dYS5IAX2EtSeqYYSBJPVAdDyEbBpLUE95aKkmrnAPIkqQBby2VpNXNh84kScAUfJ9BknOT7Eoyn+TSRZavTXJjW749yWwrf1WSO5Lc3X6+crLNl6QVou9jBknWAFcD5wGbgAuSbBqpdhHwUFWdDlwFXNnKHwB+paqeA1wIXD+phkvSStP311GcDcxX1e6qegy4Adg8UmczcF2bvhk4J0mq6s6q2tPKdwLHJFk7iYZLkiZnnDA4Gbh3aH6hlS1ap6r2Ag8D60bq/BpwZ1U9emhNlaSVq+uHzmbGqLPYictoqw9aJ8mZDC4dvXrRHSQXAxcDbNiwYYwmSdLK0/eHzhaAU4fmTwH2HKhOkhngOODBNn8K8BHgN6rqq4vtoKquqaq5qppbv3790nogSSvANDx0djtwRpKNSY4GtgBbR+psZTBADHA+cFtVVZLjgY8Cl1XVpybVaElaiXo9gNzGAC4BbgW+CNxUVTuTXJHk9a3atcC6JPPAW4B9t59eApwO/GmSz7XPMybeC0macl0/dDbOmAFVtQ3YNlJ2+dD0I8AbFlnvncA7D7ONkrQq9P6hM0nS8qqOBw0MA0nqiV6PGUiSVj7DQJJ6oOsBZMNAknqi7w+dSZKW2TQ8dCZJOhI6HEE2DCRJhoEk9YVjBpK0inX9wBkYBpLUGz50JknqlGEgSR3rwVUiw0CS+sK3lkrSKtaDEwPDQJL6wgFkSVrFvLVUkvRjPnQmSatY9+cFhoEkCcNAknrDAWRJWsV6MH5sGEhSX8TvM5Ck1at6MIRsGEiSDANJ6ppjBpKkH/NuIklSpwwDSerY3h8NrhPNPMm7iSRp1Xr0h48DcMxRazprg2EgSR17ZO+PAFg7090/yYaBJHVs35nB2hnPDCRp1Xq0nRkcc1TPzwySnJtkV5L5JJcusnxtkhvb8u1JZoeWXdbKdyV5zeSaLkkrwyPTcGaQZA1wNXAesAm4IMmmkWoXAQ9V1enAVcCVbd1NwBbgTOBc4C/b9iRJzbe/+ygAJxx7dGdtGOfM4Gxgvqp2V9VjwA3A5pE6m4Hr2vTNwDkZvHFpM3BDVT1aVV8D5tv2JEnAg99/jFvu2sPRa57Ez/300zprx8wYdU4G7h2aXwBedKA6VbU3ycPAulb+XyPrnnzIrT2IL/3Pd7nkw3eOVXfc7xtd0hPiY1Yed5tL+U7U8bc59ibHfnHWuNtcjsftV9zvaPxdL6GdS/gdrbRjeEr+nn//scElot/8hdlOby0dJwwWewpitK8HqjPOuiS5GLgYYMOGDWM06ScdM7OGZz9zCak65rMdS3kEZNzXz467zaU8mj7+Nsff6Ng1l9TOMX9Hy/LnM+42l/A7Gneby/DLXHG/o7G3N/kHs5b2d22yx/CJT13LC551Ai+cPWH8RiyDccJgATh1aP4UYM8B6iwkmQGOAx4cc12q6hrgGoC5ublD+j/k7InHcvWvn3Uoq0rSqjfOmMHtwBlJNiY5msGA8NaROluBC9v0+cBtNThH2wpsaXcbbQTOAD4zmaZLkiblCc8M2hjAJcCtwBrgA1W1M8kVwI6q2gpcC1yfZJ7BGcGWtu7OJDcBXwD2Am+uqseXqS+SpEOUpQzyHAlzc3O1Y8eOrpshSVMlyR1VNXeo6/sEsiTJMJAkGQaSJAwDSRKGgSSJHt5NlOR+4BuHsYkTgQcm1Jw+sD/9Zn/6b6X16UD9eVZVrT/UjfYuDA5Xkh2Hc3tV39iffrM//bfS+rRc/fEykSTJMJAkrcwwuKbrBkyY/ek3+9N/K61Py9KfFTdmIElaupV4ZiBJWqIVEwZJzk2yK8l8kku7bs+wJB9Icl+Se4bKnp7k40m+0n6e0MqT5D2tH3clOWtonQtb/a8kuXCo/AVJ7m7rvCfL8e0f+/fn1CT/luSLSXYm+YNp7lOSY5J8JsnnW3/e0co3Jtne2nZje4U77ZXsN7a2bU8yO7Sty1r5riSvGSo/4sdnkjVJ7kxyywrpz9fbMfG5JDta2VQec21/xye5OcmX2t+ll3Tan6qa+g+DV2t/FTgNOBr4PLCp63YNte/lwFnAPUNl7wIubdOXAle26dcCH2PwxU8vBra38qcDu9vPE9r0CW3ZZ4CXtHU+Bpy3zP05CTirTT8N+DKwaVr71Pbx1DZ9FLC9tfMmYEsrfx/w+236TcD72vQW4MY2vakde2uBje2YXNPV8Qm8BfgwcEubn/b+fB04caRsKo+5tr/rgN9p00cDx3fZn2X9wztSn9bhW4fmLwMu67pdI22cZf8w2AWc1KZPAna16fcDF4zWAy4A3j9U/v5WdhLwpaHy/eodob79M/CqldAn4CnAZxl8z/cDwMzoMcbguz1e0qZnWr2MHnf76nVxfDL4VsFPAK8Ebmntm9r+tP18nZ8Mg6k85oCfAr5GG7ftQ39WymWik4F7h+YXWlmfPbOqvgXQfj6jlR+oLwcrX1ik/IholxSez+B/01Pbp3ZJ5XPAfcDHGfzP9ztVtXeRNvy43W35w8A6lt7P5fRu4K3Aj9r8Oqa7PzD4/vR/TXJHBt+bDtN7zJ0G3A98sF3K++skx9Jhf1ZKGCx2LWxab5M6UF+WWr7skjwV+AfgD6vquwerukhZr/pUVY9X1fMY/I/6bODnD9KGXvcnyeuA+6rqjuHig7Sh1/0Z8tKqOgs4D3hzkpcfpG7f+zTD4NLxX1XV84HvM7gsdCDL3p+VEgYLwKlD86cAezpqy7i+neQkgPbzvlZ+oL4crPyURcqXVZKjGATB31bVP7biqe4TQFV9B/h3Btdlj0+y76thh9vw43a35ccx+LrXpfZzubwUeH2SrwM3MLhU9G6mtz8AVNWe9vM+4CMMQntaj7kFYKGqtrf5mxmEQ3f9We7rfEfiwyBldzMY5No3oHVm1+0aaeMs+48Z/Bn7DxS9q03/MvsPFH2mlT+dwTXGE9rna8DT27LbW919A0WvXea+BPgb4N0j5VPZJ2A9cHybfjLwSeB1wN+z/4Drm9r0m9l/wPWmNn0m+w+47mYw2NrZ8Qm8gv8fQJ7a/gDHAk8bmv5P4NxpPeba/j4JPLtNv731pbP+LPvBeKQ+DEbbv8zgWu/bum7PSNv+DvgW8EMGiX0Rg2uynwC+0n7u+wMMcHXrx93A3NB2fhuYb5/fGiqfA+5p67yXkUGpZejPyxicct4FfK59XjutfQKeC9zZ+nMPcHkrP43BHRnzDP4hXdvKj2nz8235aUPbeltr8y6G7t7o6vhk/zCY2v60tn++fXbu2+e0HnNtf88DdrTj7p8Y/GPeWX98AlmStGLGDCRJh8EwkCQZBpIkw0CShGEgScIwkCRhGEiSMAwkScD/AYBmKwmgQcFrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#find the variance of each genes of signature matrix and add a new coloum in the data frame\n",
    "\n",
    "var_signature=normalized_signature.var(axis=1)\n",
    "normalized_signature['Variance']=var_signature.values\n",
    "plt.plot(sorted(var_signature))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def threshold_data_sets(mixture_data,signature_data_with_variance,threshold):\n",
    "    \n",
    "    ''' Before using this function, it is better to plot sorted_variance values to decide threshold.\n",
    "    \n",
    "     Signature matrix should include variance values\n",
    "    \n",
    "    \n",
    "    '''\n",
    "\n",
    "    threshold_signature = signature_data_with_variance[signature_data_with_variance['Variance']>=threshold]\n",
    "    \n",
    "    threshold_signature=threshold_signature.drop(['Variance'],axis=1)\n",
    "    \n",
    "    threshold_signature=threshold_signature.fillna(0)\n",
    "    \n",
    "    gene_signature=threshold_signature.index.values\n",
    "    \n",
    "    gene_mixture=normalized_mixture.index.values\n",
    "    \n",
    "    genes=[genes for genes in gene_mixture if genes not in gene_signature]\n",
    "    \n",
    "    threshold_mixture=mixture_data.drop(index=genes)\n",
    "    threshold_mixture=threshold_mixture.fillna(0)\n",
    "    \n",
    "    return threshold_signature, threshold_mixture\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_test_data_score(samples,cell_type, gene, mixture_data,signature_matrix):\n",
    "    '''\n",
    "    This function gives the scores of randomly generated synthetic mixture data set\n",
    "       \n",
    "       sample:  number of samples for synthetic mixture data\n",
    "       cell_type: # number cell_types\n",
    "       gene: number of genes in the signature matrix\n",
    "       mixture data:  threshold mixture data\n",
    "       signature data : threshold signature matrix'''\n",
    "    \n",
    "    #generating synthetic mixture data \n",
    "    test_data=np.empty(shape=[samples,cell_type])\n",
    "    test_mixture=np.empty(shape=[samples,gene])\n",
    "    test_mixture_df=pd.DataFrame(index= mixture_data.index)\n",
    "    result_test=pd.DataFrame(index=range(samples),columns=signature_matrix.columns.values.tolist())\n",
    "    score_data=np.empty(shape=[samples,1])\n",
    "    for i in range(samples):\n",
    "        test_data[i]=np.array(np.random.random(cell_type))\n",
    "        test_data[i] /=np.sum(test_data[i])\n",
    "        test_mixture[i]=np.matmul(test_data[i],signature_matrix.T)\n",
    "        test_mixture_df[i]=pd.DataFrame(test_mixture[i],index= mixture_data.index)\n",
    "        test_mixture_df=test_mixture_df.fillna(0)\n",
    "        #using synthetic mixture data in the model:\n",
    "        result_test=Nu_SVR(test_mixture_df, signature_matrix)\n",
    "        score_data[i]=r2_score(test_data[i],result_test.iloc[i])\n",
    "    return score_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ENSG00000074800.13</th>\n",
       "      <th>ENSG00000142657.20</th>\n",
       "      <th>ENSG00000028137.18</th>\n",
       "      <th>ENSG00000142676.12</th>\n",
       "      <th>ENSG00000169442.8</th>\n",
       "      <th>ENSG00000198830.10</th>\n",
       "      <th>ENSG00000000938.12</th>\n",
       "      <th>ENSG00000130775.15</th>\n",
       "      <th>ENSG00000162511.7</th>\n",
       "      <th>ENSG00000182866.16</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000212907.2</th>\n",
       "      <th>ENSG00000198886.2</th>\n",
       "      <th>ENSG00000198786.2</th>\n",
       "      <th>ENSG00000198695.2</th>\n",
       "      <th>ENSG00000198727.2</th>\n",
       "      <th>ENSG00000210195.2</th>\n",
       "      <th>ENSG00000210196.2</th>\n",
       "      <th>ERCC-00002</th>\n",
       "      <th>ERCC-00074</th>\n",
       "      <th>ERCC-00096</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD8_naive</th>\n",
       "      <td>0.019306</td>\n",
       "      <td>0.001534</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>0.029999</td>\n",
       "      <td>0.012729</td>\n",
       "      <td>0.016410</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.000971</td>\n",
       "      <td>0.031175</td>\n",
       "      <td>0.030230</td>\n",
       "      <td>...</td>\n",
       "      <td>0.426865</td>\n",
       "      <td>0.245647</td>\n",
       "      <td>0.108467</td>\n",
       "      <td>0.075782</td>\n",
       "      <td>0.289643</td>\n",
       "      <td>0.006568</td>\n",
       "      <td>0.017030</td>\n",
       "      <td>0.083341</td>\n",
       "      <td>0.069839</td>\n",
       "      <td>0.053507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_CM</th>\n",
       "      <td>0.019509</td>\n",
       "      <td>0.001310</td>\n",
       "      <td>0.002110</td>\n",
       "      <td>0.024920</td>\n",
       "      <td>0.016353</td>\n",
       "      <td>0.017403</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>0.034476</td>\n",
       "      <td>0.031658</td>\n",
       "      <td>...</td>\n",
       "      <td>0.297374</td>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.089361</td>\n",
       "      <td>0.052855</td>\n",
       "      <td>0.232452</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.010333</td>\n",
       "      <td>0.074055</td>\n",
       "      <td>0.067231</td>\n",
       "      <td>0.045749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_EM</th>\n",
       "      <td>0.024111</td>\n",
       "      <td>0.001771</td>\n",
       "      <td>0.005592</td>\n",
       "      <td>0.011015</td>\n",
       "      <td>0.017618</td>\n",
       "      <td>0.023779</td>\n",
       "      <td>0.010812</td>\n",
       "      <td>0.001883</td>\n",
       "      <td>0.040473</td>\n",
       "      <td>0.041626</td>\n",
       "      <td>...</td>\n",
       "      <td>0.362777</td>\n",
       "      <td>0.206895</td>\n",
       "      <td>0.099897</td>\n",
       "      <td>0.086693</td>\n",
       "      <td>0.259394</td>\n",
       "      <td>0.008106</td>\n",
       "      <td>0.009930</td>\n",
       "      <td>0.148139</td>\n",
       "      <td>0.114487</td>\n",
       "      <td>0.107721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8_TE</th>\n",
       "      <td>0.014170</td>\n",
       "      <td>0.001175</td>\n",
       "      <td>0.003650</td>\n",
       "      <td>0.011725</td>\n",
       "      <td>0.016800</td>\n",
       "      <td>0.012720</td>\n",
       "      <td>0.012201</td>\n",
       "      <td>0.001134</td>\n",
       "      <td>0.023218</td>\n",
       "      <td>0.023797</td>\n",
       "      <td>...</td>\n",
       "      <td>0.254792</td>\n",
       "      <td>0.137287</td>\n",
       "      <td>0.081357</td>\n",
       "      <td>0.065394</td>\n",
       "      <td>0.212041</td>\n",
       "      <td>0.004548</td>\n",
       "      <td>0.007840</td>\n",
       "      <td>0.098993</td>\n",
       "      <td>0.079170</td>\n",
       "      <td>0.064564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAIT</th>\n",
       "      <td>0.018945</td>\n",
       "      <td>0.001052</td>\n",
       "      <td>0.001996</td>\n",
       "      <td>0.016921</td>\n",
       "      <td>0.011174</td>\n",
       "      <td>0.012615</td>\n",
       "      <td>0.001082</td>\n",
       "      <td>0.001191</td>\n",
       "      <td>0.036514</td>\n",
       "      <td>0.026527</td>\n",
       "      <td>...</td>\n",
       "      <td>0.376352</td>\n",
       "      <td>0.202431</td>\n",
       "      <td>0.105858</td>\n",
       "      <td>0.064155</td>\n",
       "      <td>0.277510</td>\n",
       "      <td>0.006607</td>\n",
       "      <td>0.009190</td>\n",
       "      <td>0.092059</td>\n",
       "      <td>0.056347</td>\n",
       "      <td>0.060531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 509 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0  ENSG00000074800.13  ENSG00000142657.20  ENSG00000028137.18  \\\n",
       "CD8_naive             0.019306            0.001534            0.000521   \n",
       "CD8_CM                0.019509            0.001310            0.002110   \n",
       "CD8_EM                0.024111            0.001771            0.005592   \n",
       "CD8_TE                0.014170            0.001175            0.003650   \n",
       "MAIT                  0.018945            0.001052            0.001996   \n",
       "\n",
       "Unnamed: 0  ENSG00000142676.12  ENSG00000169442.8  ENSG00000198830.10  \\\n",
       "CD8_naive             0.029999           0.012729            0.016410   \n",
       "CD8_CM                0.024920           0.016353            0.017403   \n",
       "CD8_EM                0.011015           0.017618            0.023779   \n",
       "CD8_TE                0.011725           0.016800            0.012720   \n",
       "MAIT                  0.016921           0.011174            0.012615   \n",
       "\n",
       "Unnamed: 0  ENSG00000000938.12  ENSG00000130775.15  ENSG00000162511.7  \\\n",
       "CD8_naive             0.000091            0.000971           0.031175   \n",
       "CD8_CM                0.000640            0.001060           0.034476   \n",
       "CD8_EM                0.010812            0.001883           0.040473   \n",
       "CD8_TE                0.012201            0.001134           0.023218   \n",
       "MAIT                  0.001082            0.001191           0.036514   \n",
       "\n",
       "Unnamed: 0  ENSG00000182866.16     ...      ENSG00000212907.2  \\\n",
       "CD8_naive             0.030230     ...               0.426865   \n",
       "CD8_CM                0.031658     ...               0.297374   \n",
       "CD8_EM                0.041626     ...               0.362777   \n",
       "CD8_TE                0.023797     ...               0.254792   \n",
       "MAIT                  0.026527     ...               0.376352   \n",
       "\n",
       "Unnamed: 0  ENSG00000198886.2  ENSG00000198786.2  ENSG00000198695.2  \\\n",
       "CD8_naive            0.245647           0.108467           0.075782   \n",
       "CD8_CM               0.158015           0.089361           0.052855   \n",
       "CD8_EM               0.206895           0.099897           0.086693   \n",
       "CD8_TE               0.137287           0.081357           0.065394   \n",
       "MAIT                 0.202431           0.105858           0.064155   \n",
       "\n",
       "Unnamed: 0  ENSG00000198727.2  ENSG00000210195.2  ENSG00000210196.2  \\\n",
       "CD8_naive            0.289643           0.006568           0.017030   \n",
       "CD8_CM               0.232452           0.005435           0.010333   \n",
       "CD8_EM               0.259394           0.008106           0.009930   \n",
       "CD8_TE               0.212041           0.004548           0.007840   \n",
       "MAIT                 0.277510           0.006607           0.009190   \n",
       "\n",
       "Unnamed: 0  ERCC-00002  ERCC-00074  ERCC-00096  \n",
       "CD8_naive     0.083341    0.069839    0.053507  \n",
       "CD8_CM        0.074055    0.067231    0.045749  \n",
       "CD8_EM        0.148139    0.114487    0.107721  \n",
       "CD8_TE        0.098993    0.079170    0.064564  \n",
       "MAIT          0.092059    0.056347    0.060531  \n",
       "\n",
       "[5 rows x 509 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''''an example of threshold_signature and threshold_mixture data'''\n",
    "\n",
    "sample_signature,sample_mixture=threshold_data_sets(normalized_mixture,normalized_signature,6.144565610406107e-05)\n",
    "sample_signature=sample_signature.T\n",
    "#save the sample data tests\n",
    "sample_signature.to_csv('sample_signature.csv') \n",
    "sample_mixture.to_csv('sample_mixture.csv') \n",
    "\n",
    "sample_signature.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7747824694401021"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'an example for using generate_test_data_score function to find average score of randomly generated mixture databy using above sample threshold mixture and signature matrix'''\n",
    "\n",
    "a=generate_test_data_score(100,29,500,sample_mixture,sample_signature)\n",
    "average_score=np.mean(a)\n",
    "average_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def variance_threshold_score(mixture_data,signature_data,lower_bound_threshold,upper_bound_threshold,sample_size,cell_number):\n",
    "    \n",
    "    ''' This function find the best threshold with min of  average score of randomly generated mixture data\n",
    "    \n",
    "    mixture_data& signature data should be normalized data\n",
    "    \n",
    "    signature_data should not include Variance values\n",
    "    \n",
    "    lower_bound_threshold is the number of highest top variance we want to check\n",
    "    \n",
    "    upper_bound_threshold is the number  of highest top variance values we do not check \n",
    "    \n",
    "    ( usually top 10 highest valuence gives negavite values so we need upper_bound_threshold to exclude them) \n",
    "\n",
    "    sample size: number of randomly samples\n",
    "    cell_type: number of cell type in the signature matrix\n",
    "    \n",
    "    '''\n",
    "    #find variance values and add them to signature matrix\n",
    "    var_mixture_signature=signature_data.var(axis=1)\n",
    "    signature_data['Variance']=var_mixture_signature.values\n",
    "    var_mixture_signature=sorted(var_mixture_signature)\n",
    "    #choose the highest variance values we want to use\n",
    "    highest_variances=var_mixture_signature[-lower_bound_threshold:-upper_bound_threshold]\n",
    "    \n",
    "    highest_variance_score=pd.DataFrame(columns=['Variance','score'])\n",
    "    \n",
    "    for var in highest_variances:\n",
    "        #\n",
    "        a,b=threshold_data_sets(mixture_data,signature_data,var)\n",
    "        number_of_genes=len(a.index)\n",
    "        x=generate_test_data_score(sample_size,cell_number,number_of_genes,b,a)\n",
    "        mean_score=np.mean(x)\n",
    "        variance_score=pd.DataFrame([[var,mean_score]],columns=['Variance','score'])\n",
    "        highest_variance_score=highest_variance_score.append(variance_score,ignore_index=True)\n",
    "        best_variance=highest_variance_score.loc[highest_variance_score['score']==highest_variance_score['score'].min(axis=0)]\n",
    "        best_variance=np.asarray(best_variance)\n",
    "    \n",
    "    return print('best variance and its score is:',best_variance),highest_variance_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best variance and its score is: [[ 0.07711138 -1.53666487]]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variance</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.077111</td>\n",
       "      <td>-1.536665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variance     score\n",
       "0  0.077111 -1.536665"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# an example with the top 2 highest variance\n",
    "x,y=variance_threshold_score(normalized_mixture,normalized_signature,2,1,50,29)\n",
    "\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best variance and its score is: [[6.32647657e-05 7.77195488e-01]]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variance</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.777195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.782918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.795478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.804750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.778866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variance     score\n",
       "0  0.000063  0.777195\n",
       "1  0.000064  0.782918\n",
       "2  0.000064  0.795478\n",
       "3  0.000064  0.804750\n",
       "4  0.000064  0.778866"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# an example with the top 500th,499th,498th,497th,496th highest variance values\n",
    "x,y=variance_threshold_score(normalized_mixture,normalized_signature,500,495,50,29)\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
